### ISSUE TITLE
Modern techniques for dependency resolution

### ISSUE THREAD

**Issue by mratsim** [nim-lang/nimble#890]
Just a recap of a short discussion with @haxscramper and @FedericoCeratto on modern dependency resolution.

Nimble found a certain combination of dependency impossible to resolve while there is actually a solution.

Modern package managers that handles tens of thousands of packages and/or multiversioning are gradually delegating dependency resolution to SAT solvers as this is a problem that can be modelled as a Satisfiability problem.

Writeups:
- Anaconda, the scientific python distribution which manages 1às of thousands of branches, versions, architectures, optional dependencies: https://www.anaconda.com/blog/understanding-and-improving-condas-performance. Note for personally using for data science, it is still way too slow.
- Mamba, a package manager for conda in C++ that is significantly faster: https://wolfv.medium.com/making-conda-fast-again-4da4debfb3b7

TLDR;
- Both are looking into SAT solver, conda was looking into using Z3 specifically https://github.com/Z3Prover/z3/issues/211.
If we provide Z3, for DrNim nimble can piggyback on it.
- Mamba choose a pragmatic solution `libsolv` from the OpenSuse project which is very package management oriented: https://github.com/openSUSE/libsolv

Note: I think there are other approaches possible from writing that ourselves or using alternative theories like Integer Linear Programming (ILP) but it's more pragmatic to reuse existing blocks that were optimized and stabilized over years of usage here and regarding ILP, I tried to wrap the one for LLVM (http://isl.gforge.inria.fr/) or write one myself (https://github.com/mratsim/hydra) but those are really complex.

**Comment by FedericoCeratto** [comment]
Datapoints regarding distributing Z3: it is packaged in most distributions, it appears to be stable, without concerning dependencies, and runs on most architectures.

Then a valid solution is impossible, the dependency solver in apt/aptitude can suggest multiple alternative solutions that break/ignore one or more constraint each, starting with the less "aggressive" first. Z3 might help doing the same easily.

Also, https://ci.debian.net/doc/ run automated tests against multiple solutions and across multiple packages to spot incorrect dependency constraints. Z3 could help doing local compilation tests against combinations of dependencies to spot similar issues.

**Comment by haxscramper** [comment]
Current implementation of dependency resolution does not construct explicit dependency graph, and instead just loops https://github.com/nim-lang/nimble/blob/95e6870f60655e81ff488779c7f589fe649061ec/src/nimble.nim#L64 though requirements, almost immediately installing them https://github.com/nim-lang/nimble/blob/95e6870f60655e81ff488779c7f589fe649061ec/src/nimble.nim#L84-L86 which I believe to be the source of such bugs as https://github.com/nim-lang/nimble/issues/505 and https://github.com/nim-lang/nimble/issues/887 (could be prevented with explicit dependency graph construction).

Maybe we could adopt an approach similar to https://hal.archives-ouvertes.fr/hal-00149566/document for mapping dependency requirements to something that is fed into `z3`

Related issues:

- https://github.com/disruptek/golden/issues/21  Can't install with nimble: nimterop versions 
- https://github.com/nim-lang/nimble/issues/505  Dependency resolution depends on the order in `requires`
- https://github.com/nim-lang/nimble/issues/766  Installing package fails to find the dependent package even though it exists.
- https://github.com/nim-lang/nimble/issues/887  nimble loops infinitely trying to install codependent requirements

**Comment by stefantalpalaru** [comment]
> Both are looking into SAT solver

For good reasons: https://research.swtch.com/version-sat

**Comment by dom96** [comment]
> For good reasons: https://research.swtch.com/version-sat

This is a classic, if you haven't already, check out the "Alternatives?" section. Some of the suggestions may sound familiar.

![image](https://user-images.githubusercontent.com/246651/107438410-555db300-6b28-11eb-8088-cb2003cc9258.png)

Specifying a concrete version in a Nimble file has always been a bad idea and largely discouraged. I think that once lock files are merged into Nimble that we should look into disallowing concrete versions in "requires" completely (without lock files this is the only way you can hope to get reproducible builds).

Nim should also grow the ability to compile a package with a dependency on two different versions of the same package. I think I may have finally convinced Araq that this is necessary :)

**Comment by haxscramper** [comment]
Approximate current stats for nimble package requirement specifications:

```
total package count:      1565
processing ok:            1356
not using github:         51
http error when getting:  105
configuration parse fail: 27

# Total statistics for all `requires` patterns encountered in processed packages

verLater        (   > V    ): 25
verEarlier      (   < V    ): 4
verEqLater      (   >= V   ): 1727
verEqEarlier    (   <= V   ): 2
verIntersect    (> V & < V ): 37
verEq           (    V     ): 24
verAny          (    *     ): 315
verSpecial      (  #head   ): 35
```

**Comment by haxscramper** [comment]
I think it should be explicitly noted that while the addition of SAT solver is debatable, the dependency resolution step of nimble should be refactored into explicit dependency graph construction with subsequent full download of packages to be installed. Current solution of resolving-packages-while-installing-packages-wile-resolving packages leads to overly verbose output where each dependency can be repeated, makes resolution dependent on order of `requires` statements, and makes it hardly possible to do things like `nimble build` and `nimble test` without installing anything. 

Package can appear multiple times in the output if it is required by more than one package, but each additional encounter will have useless 'requirement already satisfied' message, cluttering output and making it much harder to determine what went wrong. 

It is possible to get all necessary dependency information by parsing `.nimble` file, but without evaluating it, and while this might seem like an overly fragile approach it worked successfully for approximately 96% of the [processed] packages. It is not possible to account for all possible ways of writing package specification of course, but things like

```nim
  version       = pkgVersion # fae

  # paravim
  installExt    = @[
    "nim", "txt", "ttf", "glsl", "c", "h",
    when defined(windows):
      "dll"
    elif defined(macosx):
      "dylib"
    elif defined(linux):
      "so"
  ]

  # metar
  include metar/version

  version = metarVersion

  # plz
  version     = CompileDate.replace("-", ".")
```
Are extremely rare, and for these cases fully downloading packages before resolving it is the only solution (as well as non-github hostings (account for 3% of all packages).

**Comment by haxscramper** [comment]
Related nim forum discussion https://forum.nim-lang.org/t/7671
